source/opt/vector_dce.cpp

```diff
@@ -0,0 +1,287 @@
+  WorkListItem second_operand;
+  second_operand.instruction =
+      def_use_mgr->GetDef(current_item.instruction->GetSingleWordInOperand(1));
+
+  analysis::TypeManager* type_mgr = context()->get_type_mgr();
+  analysis::Vector* first_type =
+      type_mgr->GetType(first_operand.instruction->type_id())->AsVector();
+  uint32_t size_of_first_operand = first_type->element_count();
+
+  for (uint32_t in_op = 2; in_op < current_item.instruction->NumInOperands();
+       ++in_op) {
+    uint32_t index = current_item.instruction->GetSingleWordInOperand(in_op);
+    if (current_item.components.Get(in_op - 2)) {
+      if (index < size_of_first_operand) {
+        first_operand.components.Set(index);
+      } else {
+        second_operand.components.Set(index - size_of_first_operand);
+      }
+    }
+  }
+
+  AddItemToWorkListIfNeeded(first_operand, live_components, work_list);
+  AddItemToWorkListIfNeeded(second_operand, live_components, work_list);
+}
+
+void VectorDCE::MarkUsesAsLive(
+    ir::Instruction* current_inst, const utils::BitVector& live_elements,
+    LiveComponentMap* live_components,
+    std::vector<VectorDCE::WorkListItem>* work_list) {
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+
+  current_inst->ForEachInId([&work_list, &live_elements, this, live_components,
+                             def_use_mgr](uint32_t* operand_id) {
+    ir::Instruction* operand_inst = def_use_mgr->GetDef(*operand_id);
+
+    if (HasVectorResult(operand_inst)) {
+      WorkListItem new_item;
+      new_item.instruction = operand_inst;
+      new_item.components = live_elements;
+      AddItemToWorkListIfNeeded(new_item, live_components, work_list);
+    }
+  });
+}
+
+bool VectorDCE::HasVectorResult(const ir::Instruction* inst) const {
+  analysis::TypeManager* type_mgr = context()->get_type_mgr();
+  if (inst->type_id() == 0) {
+    return false;
+  }
+
+  const analysis::Type* current_type = type_mgr->GetType(inst->type_id());
+  bool is_vector = current_type->AsVector() != nullptr;
+  return is_vector;
+}
+
+bool VectorDCE::RewriteInstructions(
+    ir::Function* function,
+    const VectorDCE::LiveComponentMap& live_components) {
+  bool modified = false;
+  function->ForEachInst(
+      [&modified, this, live_components](ir::Instruction* current_inst) {
+        if (!context()->IsCombinatorInstruction(current_inst)) {
+          return;
+        }
+
+        auto live_component = live_components.find(current_inst->result_id());
+        if (live_component == live_components.end()) {
+          // If this instruction is not in live_components then it does not
+          // produce a vector, or it is never referenced and ADCE will remove
+          // it.  No point in trying to differentiate.
+          return;
+        }
+
+        // If no element in the current instruction is used replace it with an
+        // OpUndef.
+        if (live_component->second.Empty()) {
```

Comment:
Maybe just refactor Type2Undef from MemPass and generate a reusable undef? Then you can just replace all uses and not have special cases for OpPhi.

Reply:
I can can generate a global OpUndef and reuse the same one.  I did not do that because I wanted to reduce the amount if time spent updating the def-use chains.  With this change, there are very few changes to the def-use information.

I will change it.  I think it will generate cleaner code.  However, I won't factor out the stuff from mempass.  I don't want to muddy this PR.  We can do that on its own later.